snappy/desktop.go

```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snappy
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/helpers"
+)
+
+// this is a bit confusing, a localestring in xdg is:
+// "key=value" or "key[locale]=" with locale as:
+//  lang_COUNTRY@MODFIER (or a subset of this)
+var desktopFileI18nPattern = `(|\[[a-zA-Z_@]+\])`
+var validDesktopFileLines = []*regexp.Regexp{
+	// headers
+	regexp.MustCompile(`^\[Desktop Entry\]$`),
+	regexp.MustCompile(`^\[Desktop Action [a-zA-Z0-9]+\]$`),
```

Comment:
Interesting. The way I read https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s10.html#extra-actions-keys 
```
Each action is identified by a string, following the same format as key names (see the section called “Entries”). Each identifier is associated with an action group that must be present in the .desktop file. The action group is a group named Desktop Action %s, where %s is the action identifier. 
```
and in https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s02.html#entries: ` Only the characters A-Za-z0-9- may be used in key names. `

However this may actually not reflect reality. I see for example `[Desktop Action window-shot]` from gnome-screenshot which clearly violates this :) So maybe this needs to be relaxed for the real world.